packages/mattermost-redux/src/reducers/entities/apps.ts

```diff
@@ -3,25 +3,81 @@
 import {combineReducers} from 'redux';
 
 import {AppsTypes} from 'mattermost-redux/action_types';
-import {AppBinding, AppsState} from 'mattermost-redux/types/apps';
+import {AppBinding, AppCommandFormMap, AppsState} from 'mattermost-redux/types/apps';
 import {GenericAction} from 'mattermost-redux/types/actions';
 import {validateBindings} from 'mattermost-redux/utils/apps';
 
 // This file's contents belong to the Apps Framework feature.
 // Apps Framework feature is experimental, and the contents of this file are
 // susceptible to breaking changes without pushing the major version of this package.
 
-export function bindings(state: AppBinding[] = [], action: GenericAction): AppBinding[] {
+export function mainBindings(state: AppBinding[] = [], action: GenericAction): AppBinding[] {
     switch (action.type) {
     case AppsTypes.RECEIVED_APP_BINDINGS: {
-        validateBindings(action.data);
-        return action.data || [];
+        const bindings = action.data;
+        validateBindings(bindings);
+
+        return bindings;
     }
     default:
         return state;
     }
 }
 
+function mainForms(state: AppCommandFormMap = {}, action: GenericAction): AppCommandFormMap {
+    switch (action.type) {
+    case AppsTypes.RECEIVED_APP_BINDINGS:
+        return {};
+    case AppsTypes.RECEIVED_APP_FORM: {
+        const {form, location} = action.data;
+        const newState = JSON.parse(JSON.stringify(state));
```

Comment:
Could we skip cloning parts of the state that aren't changing?
```javascript
const newState = {
    ...state,
    [location]: form,
};
```
should work here since only `state[location]` and `state` itself need to change